diff --git a/habitat/core/vector_env.py b/habitat/core/vector_env.py
index e691f8eb8..28a81f7a8 100644
--- a/habitat/core/vector_env.py
+++ b/habitat/core/vector_env.py
@@ -54,6 +54,7 @@ RESET_COMMAND = "reset"
 RENDER_COMMAND = "render"
 CLOSE_COMMAND = "close"
 CALL_COMMAND = "call"
+CALL_SIM_COMMAND = "call_sim"
 COUNT_EPISODES_COMMAND = "count_episodes"
 
 EPISODE_OVER_NAME = "episode_over"
@@ -278,6 +279,24 @@ class VectorEnv:
                         result = result_or_fn
 
                     connection_write_fn(result)
+                elif command == CALL_SIM_COMMAND:
+                    function_name, function_args = data
+                    if function_args is None:
+                        function_args = {}
+
+                    try:
+                        result_or_fn = getattr(env._env._sim, function_name)
+
+                        if len(function_args) > 0 or callable(result_or_fn):
+                            result = result_or_fn(**function_args)
+                        else:
+                            result = result_or_fn
+
+                        connection_write_fn(result)
+                    except Exception as e:
+                        print(e)
+                        import traceback
+                        traceback.print_exc()
 
                 elif command == COUNT_EPISODES_COMMAND:
                     connection_write_fn(len(env.episodes))
@@ -518,6 +537,18 @@ class VectorEnv:
         result = self._connection_read_fns[index]()
         return result
 
+    def call_sim_at(
+        self,
+        index: int,
+        function_name: str,
+        function_args: Optional[Dict[str, Any]] = None,
+    ) -> Any:
+        self._connection_write_fns[index](
+            (CALL_SIM_COMMAND, (function_name, function_args))
+        )
+        result = self._connection_read_fns[index]()
+        return result
+
     def call(
         self,
         function_names: List[str],
diff --git a/habitat/utils/visualizations/utils.py b/habitat/utils/visualizations/utils.py
index 413cebd74..028556aca 100644
--- a/habitat/utils/visualizations/utils.py
+++ b/habitat/utils/visualizations/utils.py
@@ -203,7 +203,9 @@ def tile_images(render_obs_images: List[np.ndarray]) -> np.ndarray:
     return final_im
 
 
-def observations_to_image(observation: Dict, info: Dict) -> np.ndarray:
+def observations_to_image(
+    observation: Dict, info: Dict, extra_sensors
+) -> np.ndarray:
     r"""Generate image of single frame from observation and info
     returned from a single environment step().
 
@@ -230,6 +232,18 @@ def observations_to_image(observation: Dict, info: Dict) -> np.ndarray:
             depth_map = depth_map.astype(np.uint8)
             depth_map = np.stack([depth_map for _ in range(3)], axis=2)
             render_obs_images.append(depth_map)
+        elif sensor_name in extra_sensors:
+            sensor_data = observation[sensor_name]
+            if not isinstance(sensor_data, np.ndarray):
+                sensor_data = sensor_data.cpu().numpy()
+
+            if sensor_data.shape[-1] == 1:
+                sensor_data = np.repeat(sensor_data, 3, axis=2)
+
+            # Resize to match height of rgb image:
+            ratio = observation["rgb"].shape[0] / sensor_data.shape[0]
+            sensor_data = cv2.resize(sensor_data, None, fx=ratio, fy=ratio)
+            render_obs_images.append(sensor_data)
 
     # add image goal if observation has image_goal info
     if "imagegoal" in observation:
diff --git a/habitat_baselines/common/tensorboard_utils.py b/habitat_baselines/common/tensorboard_utils.py
index 6b41e763e..9eef90945 100644
--- a/habitat_baselines/common/tensorboard_utils.py
+++ b/habitat_baselines/common/tensorboard_utils.py
@@ -12,6 +12,17 @@ import torch
 from torch.utils.tensorboard import SummaryWriter
 
 try:
+    import importlib.metadata
+
+    # Quick fix for packages missing metadata...
+    _original_distributions = importlib.metadata.distributions
+    def patched_distributions(**kwargs):
+        for d in _original_distributions(**kwargs):
+            if d.metadata is not None:
+                yield d
+
+    importlib.metadata.distributions = patched_distributions
+
     import wandb
 except ImportError:
     wandb = None
@@ -101,6 +112,8 @@ class WeightsAndBiasesWriter:
             wb_kwargs["entity"] = config.WB.ENTITY
         if config.WB.GROUP != "":
             wb_kwargs["group"] = config.WB.GROUP
+        if config.WB.MODE != "":
+            wb_kwargs["mode"] = config.WB.MODE
         slurm_info_dict = {
             k[len("SLURM_") :]: v
             for k, v in os.environ.items()
@@ -110,7 +123,7 @@ class WeightsAndBiasesWriter:
             raise ValueError(
                 "Requested to log with wandb, but wandb is not installed."
             )
-
+        print(f"{wb_kwargs=}")
         self.run = wandb.init(
             config={"slurm": slurm_info_dict, **config}, **wb_kwargs
         )
diff --git a/habitat_baselines/rl/models/rnn_state_encoder.py b/habitat_baselines/rl/models/rnn_state_encoder.py
index f2c8be21e..9c0411386 100644
--- a/habitat_baselines/rl/models/rnn_state_encoder.py
+++ b/habitat_baselines/rl/models/rnn_state_encoder.py
@@ -290,7 +290,7 @@ class RNNStateEncoder(nn.Module):
         )
 
         x, hidden_states = self.rnn(
-            x.unsqueeze(0), self.unpack_hidden(hidden_states)
+            x.unsqueeze(0), self.unpack_hidden(hidden_states).contiguous()
         )
         hidden_states = self.pack_hidden(hidden_states)
 
